nixos/modules/services/networking/networkmanager.nix

```diff
@@ -492,6 +539,30 @@ in {
       aliases = [ "dbus-org.freedesktop.nm-dispatcher.service" ];
     };
 
+    systemd.services.NetworkManager-ensure-profiles = mkIf (cfg.profiles != { }) {
+      description = "This unit creates network-manager profiles";
+      wantedBy = [ "multi-user.target" ];
+      before = [ "network-online.target" ];
+      script = let
+        path = id: "/run/NetworkManager/system-connections/${id}.nmconnection";
+      in ''
+        mkdir -p /run/NetworkManager/system-connections
+      '' + lib.concatMapStringsSep "\n"
+        (profile: ''
+          ${pkgs.envsubst}/bin/envsubst -i ${ini.generate (lib.escapeShellArg profile.n) profile.v} > ${path (lib.escapeShellArg profile.n)}
+          chmod 0400 ${path (lib.escapeShellArg profile.n)}
+        '') (lib.mapAttrsToList (n: v: { inherit n v; }) cfg.profiles)
+      + ''
+        if systemctl is-active --quiet NetworkManager; then
+          ${pkgs.networkmanager}/bin/nmcli connection reload
+        fi
+      '';
+      serviceConfig = {
+        EnvironmentFile = cfg.environmentFiles;
```

Comment:
In theory you could instead use the systemd LoadCreds functionality and then maybe don’t load the secrets into the environment for the whole unit but just for envsubst. From my understanding that would be current best-practice.
In practice though this unit is very lightweight (thus little code to compromise the environment) and you are writing the secrets to a root readable position anyways … So this suggestion is probably overkill.

Reply:
I'll give this a look, never heard of LoadCreds :see_no_evil: sounds interesting, I'll add it if it doesn't add too much complexity.

Reply:
If you mean https://freedesktop.org/software/systemd/man/systemd-creds.html with LoadCreds then it will probably make the module more complicated then it's worth.